umpost/um2netcdf.py

```diff
@@ -92,46 +102,194 @@ def convert_proleptic(time):
+
+    Parameters
+    ----------
+    longitude: (array) 1D array of longitude grid points.
+    grid_type: (string) model horizontal grid type.
+    dlon: (float) zonal spacing between longitude grid points. 
+    """
+    min_longitude = longitude_points[0]
+    min_longitude_u_ND = 0.5*dlon
+
+    return (
+        (min_longitude == 0 and grid_type == GRID_END_GAME) or
+        (
+            np.allclose(min_longitude_u_ND, min_longitude) and
+            grid_type == GRID_NEW_DYNAMICS
+        )
+    )
+
+
+def add_latlon_coord_bounds(cube_coordinate):
+    """
+    Add bounds to horizontal coordinate (longitude or latitude) if
+    they don't already exist. Edits coordinate in place.
+
+    Parameters
+    ----------
+    cube_coordinate: coordinate object from iris cube.
+    """
+    coordinate_name = cube_coordinate.name()
+    if coordinate_name not in [LON_COORD_NAME, LAT_COORD_NAME]:
+        raise ValueError(
+                f"Wrong coordinate {coordinate_name} supplied. "
+                f"Expected one of {LON_COORD_NAME}, {LAT_COORD_NAME}."
+            )
+
+    if cube_coordinate.has_bounds(): 
+        # Don't change bounds if already exist
+        return
+    
+    if len(cube_coordinate.points) > 1:
+        cube_coordinate.guess_bounds()
     else:
-        lon.var_name = 'lon'
+        # For length 1, assume it's global. guess_bounds doesn't work in this case
+        cube_coordinate.bounds = GLOBAL_COORD_BOUNDS[coordinate_name]
+
+
+
+def fix_latlon_coords(cube, grid_type, dlat, dlon):
+    """
+    Wrapper function to modify cube's horizontal coordinates 
+    (latitude and longitude). Converts to float64, adds grid bounds,
+    and renames coordinates. Modifies cube in place.
+
+    Parameters
+    ----------
+    cube: Iris cube object (modified in place).
+    grid_type: (string) model horizontal grid type.
+    dlat: (float) meridional spacing between latitude grid points.
+    NB - Only applies to variables on the main horizontal grids, 
+    and not the river grid.
+    dlon: (float) zonal spacing between longitude grid points. 
+    NB - Only applies to variables on the main horizontal grids, 
+    and not the river grid.
+    """
+
+    # TODO: Check that we don't need the double coordinate lookup.
+    try:
+        latitude_coordinate = cube.coord(LAT_COORD_NAME)
+        longitude_coordinate = cube.coord(LON_COORD_NAME)
+        # Force to double for consistency with CMOR
+        latitude_coordinate.points = latitude_coordinate.points.astype(np.float64)
+        longitude_coordinate.points = longitude_coordinate.points.astype(np.float64)
+        
+        add_latlon_coord_bounds(latitude_coordinate)
+        add_latlon_coord_bounds(longitude_coordinate)
+        
+        fix_lat_coord_name(latitude_coordinate, grid_type, dlat)
+        fix_lon_coord_name(longitude_coordinate, grid_type, dlon)
+        
+    except iris.exceptions.CoordinateNotFoundError:
```

Comment:
If the `cube.coord(name)` lookups are the only calls where `CoordinateNotFoundError` can occur, the `except` block can be moved to cover those 2 lines.

Reply:
This looks good as the try/except is really small & easy to follow.

Also, the `TODO` looks like it's left over from an earlier discussion.

Reply:
Ah yep, I've removed that leftover `TODO`